wallet-new/src/Cardano/Wallet/Kernel/Keystore.hs

```diff
@@ -0,0 +1,194 @@
+import           System.Wlog (CanLog (..), HasLoggerName (..), LoggerName (..),
+                     logMessage)
+
+import           Cardano.Wallet.Kernel.DB.HdWallet (eskToHdRootId)
+import           Cardano.Wallet.Kernel.Types (WalletId (..))
+
+data StorageState =
+      StorageReleased
+      -- ^ The internal storage was released
+    | StorageInitialised
+      -- ^ The internal storage is initialised
+
+-- Internal storage necessary to smooth out the legacy 'UserSecret' API.
+data InternalStorage = InternalStorage {
+      _storageSecret :: !UserSecret
+    , _storageState  :: !StorageState
+    }
+
+data Keystore = Keystore (MVar InternalStorage)
+
+-- | Internal monad used to smooth out the 'WithLogger' dependency imposed
+-- by 'Pos.Util.UserSecret', to not commit to any way of logging things just yet.
+newtype KeystoreM a = KeystoreM { fromKeystore :: IdentityT IO a }
+                    deriving (Functor, Applicative, Monad, MonadIO)
+
+instance HasLoggerName KeystoreM where
+    askLoggerName = return (LoggerName "Keystore")
+    modifyLoggerName _ action = action
+
+instance CanLog KeystoreM where
+    dispatchMessage _ln sev txt = logMessage sev txt
+
+{-------------------------------------------------------------------------------
+  Creating a keystore
+-------------------------------------------------------------------------------}
+
+-- | Creates a new keystore.
+newKeystore :: MonadIO m => FilePath -> m Keystore
+newKeystore fp = liftIO $ runIdentityT $ fromKeystore $ do
+    us <- takeUserSecret fp
+    Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+-- | Creates a legacy 'Keystore' by reading the 'UserSecret' from a 'NodeContext'.
+-- Hopefully this function will go in the near future.
+newLegacyKeystore :: MonadIO m => NodeContext -> m Keystore
+newLegacyKeystore ctx = do
+     us <- atomically $ readTVar $ ncUserSecret ctx
+     Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+-- | Creates a 'Keystore' out of a randomly generated temporary file (i.e.
+-- inside your $TMPDIR of choice). Suitable for testing.
+newInMemoryTransientKeystore :: MonadIO m => m Keystore
+newInMemoryTransientKeystore = liftIO $ runIdentityT $ fromKeystore $ do
+    tempDir         <- liftIO getTemporaryDirectory
+    (tempFile, hdl) <- liftIO $ openTempFile tempDir "keystore.key"
+    liftIO $ hClose hdl
+    us <- takeUserSecret tempFile
+    Keystore <$> newMVar (InternalStorage us StorageInitialised)
+
+
+-- | Release the resources associated with this 'Keystore'.
+-- This function is idempotent and can be called multiple times.
+releaseKeystore :: MonadIO m => Keystore -> m ()
+releaseKeystore (Keystore ks) = liftIO $ modifyMVar ks $ \(InternalStorage us st) ->
+    case st of
+         StorageReleased      -> return ((InternalStorage us StorageReleased), ())
+         StorageInitialised   -> do
+             writeUserSecretRelease us
+             return ((InternalStorage us StorageReleased), ())
```

Comment:
If would be nicer to export a `bracketKeystore` instead.

Reply:
Indeed, if it is true that _when_ the wallet is initialized using the legacy keystore, our release process here should be different, then it's even more important to only export `bracket` style functions, to guarantee that we do the right thing.